rust/agama-server/src/cert.rs

```diff
@@ -1,84 +1,108 @@
-pub fn create_certificate() -> Result<(X509, PKey<Private>), ErrorStack> {
-    let rsa = Rsa::generate(2048)?;
-    let key = PKey::from_rsa(rsa)?;
-
-    let mut x509_name = X509NameBuilder::new()?;
-    x509_name.append_entry_by_text("O", "Agama")?;
-    x509_name.append_entry_by_text("CN", "localhost")?;
-    let x509_name = x509_name.build();
-
-    let mut builder = X509::builder()?;
-    builder.set_version(2)?;
-    let serial_number = {
-        let mut serial = BigNum::new()?;
-        serial.rand(159, MsbOption::MAYBE_ZERO, false)?;
-        serial.to_asn1_integer()?
-    };
-    builder.set_serial_number(&serial_number)?;
-    builder.set_subject_name(&x509_name)?;
-    builder.set_issuer_name(&x509_name)?;
-    builder.set_pubkey(&key)?;
-
-    let not_before = Asn1Time::days_from_now(0)?;
-    builder.set_not_before(&not_before)?;
-    let not_after = Asn1Time::days_from_now(365)?;
-    builder.set_not_after(&not_after)?;
-
-    builder.append_extension(BasicConstraints::new().critical().ca().build()?)?;
-
-    builder.append_extension(
-        SubjectAlternativeName::new()
-            // use the default Agama host name
-            // TODO: use the gethostname crate and use the current real hostname
-            .dns("agama")
-            // use the default name for the mDNS/Avahi
-            // TODO: check which name is actually used by mDNS, to avoid
-            // conflicts it might actually use something like agama-2.local
-            .dns("agama.local")
-            .build(&builder.x509v3_context(None, None))?,
-    )?;
-
-    let subject_key_identifier =
-        SubjectKeyIdentifier::new().build(&builder.x509v3_context(None, None))?;
-    builder.append_extension(subject_key_identifier)?;
-
-    builder.sign(&key, MessageDigest::sha256())?;
-    let cert = builder.build();
-
-    // for debugging you might dump the certificate to a file:
-    // use std::io::Write;
-    // let mut cert_file = std::fs::File::create("agama_cert.pem").unwrap();
-    // let mut key_file = std::fs::File::create("agama_key.pem").unwrap();
-    // cert_file.write_all(cert.to_pem().unwrap().as_ref()).unwrap();
-    // key_file.write_all(key.private_key_to_pem_pkcs8().unwrap().as_ref()).unwrap();
-
-    Ok((cert, key))
+const DEFAULT_CERT_DIR: &str = "/etc/agama.d/ssl";
+
+pub struct Certificate {
+    pub cert: X509,
+    pub key: PKey<Private>,
+}
+
+impl Certificate {
+    /// Writes cert, key to (for now well known) location(s)
+    pub fn write(&self) -> anyhow::Result<()> {
+        // check and create default dir if needed
+        if !Path::new(DEFAULT_CERT_DIR).is_dir() {
+            std::fs::create_dir_all(DEFAULT_CERT_DIR)?;
+        }
+
+        if let Ok(bytes) = self.cert.to_pem() {
+            fs::write(Path::new(DEFAULT_CERT_DIR).join("cert.pem"), bytes)?;
+        }
+        if let Ok(bytes) = self.key.private_key_to_pem_pkcs8() {
+            fs::write(Path::new(DEFAULT_CERT_DIR).join("key.pem"), bytes)?;
```

Comment:
This will leave the key file world-readable.
It seems we should use https://doc.rust-lang.org/std/os/unix/fs/trait.OpenOptionsExt.html#tymethod.mode to set the file mode before writing to it

Reply:
do we care for self generated certificate for installer which becomes useless once the installation is over?

Reply:
I agree with @mvidner in this case. I will not hurt to set the correct permissions in this case. We are already doing that for the token so, why not?

Reply:
Ah, thanks for the tip, this method shows how it is done: https://github.com/openSUSE/agama/blob/8bebd79d8990ec690ef76ca8575c086b3044b654/rust/agama-lib/src/auth.rs#L98